services/cd-service/pkg/repository/transformer.go

```diff
@@ -356,68 +416,67 @@ type CleanupOldApplicationVersions struct {
 	Application string
 }
 
-func (c *CleanupOldApplicationVersions) Transform(ctx context.Context, fs billy.Filesystem) (string, error) {
+// Finds old releases for an application
+func findOldApplicationVersions(fs billy.Filesystem, name string) ([]uint64, error) {
 	var state = &State{Filesystem: fs}
 	// 1) get release in each env:
 	envConfigs, err := state.GetEnvironmentConfigs()
 	if err != nil {
-		return "", fmt.Errorf("CleanupOldApplicationVersions: error accessing env configs: %w", err)
+		return nil, err
 	}
-	var oldestDeployedVersion *uint64 = nil
+	versions, err := state.GetApplicationReleases(name)
+	if err != nil {
+		return nil, err
+	}
+	if len(versions) == 0 {
+		return nil, err
+	}
+	sort.Slice(versions, func(i, j int) bool {
+		return versions[i] < versions[j]
+	})
+	// Use the latest version as oldest deployed version
+	oldestDeployedVersion := versions[len(versions)-1]
 	for env := range envConfigs {
-		version, err := state.GetEnvironmentApplicationVersion(env, c.Application)
+		version, err := state.GetEnvironmentApplicationVersion(env, name)
 		if err != nil {
-			return "", fmt.Errorf("CleanupOldApplicationVersions: Could not get version for env %s app %s: %w",
-				env, c.Application, err)
-		}
-		if version != nil && oldestDeployedVersion == nil {
-			oldestDeployedVersion = version
-		} else if version != nil && oldestDeployedVersion != nil {
-			if *version < *oldestDeployedVersion {
-				oldestDeployedVersion = version
+			return nil, err
+		}
+		if version != nil {
+			if *version < oldestDeployedVersion {
+				oldestDeployedVersion = *version
 			}
 		}
 	}
+	positionOfOldestVersion := sort.Search(len(versions), func(i int) bool {
+		return versions[i] >= oldestDeployedVersion
+	})
 
-	// 2) decide if there is something to remove
-	if oldestDeployedVersion == nil {
-		return fmt.Sprintf("(no old versions of app %s were found)", c.Application), nil
-	}
-
-	if *oldestDeployedVersion <= keptVersionsOnCleanup {
-		return fmt.Sprintf("(Oldest version (%d) of app %s is still too young (<=%d) for cleanup)",
-			*oldestDeployedVersion, c.Application, keptVersionsOnCleanup), nil
+	if positionOfOldestVersion < (keptVersionsOnCleanup - 1) {
+		return nil, nil
 	}
+	return versions[0 : positionOfOldestVersion-(keptVersionsOnCleanup-1)], err
+}
 
-	// 3) remove all releases older than x
-	versions, err := state.GetApplicationReleases(c.Application)
+func (c *CleanupOldApplicationVersions) Transform(ctx context.Context, fs billy.Filesystem) (string, error) {
+	oldVersions, err := findOldApplicationVersions(fs, c.Application)
 	if err != nil {
-		return "", fmt.Errorf("cleanup: could not get application releases for app '%s': %w", c.Application, err)
+		return "", err
```

Comment:
Why remove this error message?
I would suggest to keep it.
```suggestion
		return "", fmt.Errorf("cleanup: could not get application releases for app '%s': %w", c.Application, err)
```